src/snowflake/snowpark/column.py

```diff
@@ -244,12 +251,22 @@ def __init__(
             if expr1 == "*":
                 self._expression = Star([])
             else:
-                self._expression = UnresolvedAttribute(quote_name(expr1))
+                col_name = quote_name(expr1)
+                self._expression = UnresolvedAttribute(col_name)
+                self._ast.sp_column.name = col_name
+
         elif isinstance(expr1, Expression):
             self._expression = expr1
         else:  # pragma: no cover
             raise TypeError("Column constructor only accepts str or expression.")
 
+    @property
+    def ast(self) -> proto.SpColumnExpr:
```

Comment:
This shouldn't be necessary. This logic should happen at construction time for the appropriate case. We shouldn't be introducing imperative mutations with a hard-to-follow control flow in an API without mutable state like this one. This is why I'm suggesting having clear logic in the constructor that immediately constructs the required value without mutations spread across multiple branches and functions.

Reply:
Would it be okay to still store a constructed `_ast` member and retrieve it later?

I'm facing an issue with deciding how to handle expressions of the form `df.select(df.col("A"), col("B"))`, where the provided `Column` instances are constructed before `df.select` has an opportunity to construct the necessary `SpColumnExpr` instance and populate the AST fields.

More generally any Dataframe API that accepts previously constructed Column instances will face this issue, hence the current solution. I agree with the point you made for line 1137 though as that is a clear opportunity to avoid copying from a potentially complex message into the Session AstBatch.

Reply:
Yes, I think you're right and we have to store a pre-constructed value in `_ast`. This property shouldn't be necessary though. Maybe a helper function to copy `_ast` into whatever builder is passed in?

Reply:
Sounds good, would the goal of the helper function be to replace `protobuf.message.CopyFrom(col._ast)` and `append(col._ast)` when constructing the AST one level up from the constructor?

I was thinking as long as `_ast` is ensured to be an instance of `SpColumnExpr` a helper function might not be necessary, unless there is another purpose I'm missing, please let me know!

Reply:
Yes. Definitely better without a helper if the code seems simple.